src/froll.c

```diff
@@ -397,11 +416,207 @@ void frollsumExact(double *x, uint64_t nx, ans_t *ans, int k, double fill, bool
+          iw = i-k; w = R_NegInf;
+          wmax(x, i, k, &w, &iw, true); cmax++;
+        } else if (x[i] >= w) {
+          iw = i; w = x[iw];
+        }
+        ans->dbl_v[i] = w;
+      }
+    }
+    if (truehasnf) {
+      for (; i<nx; i++) { // this loop continues from where "maybe no NAs" loop left
+        if (ISNAN(x[i])) {
+          if (ISNA(x[i])) {
+            iw = i; w = NA_REAL;
+          } else if (ISNA(w)) {
+            // do nothing because w > x[i]: NA > NaN
+          } else { // no NA in window so NaN >= than any non-NA
+            iw = i; w = R_NaN;
+          }
+        } else if (iw+k <= i) { // max left current window
+          iw = i-k; w = R_NegInf;
+          wmax(x, i, k, &w, &iw, false); cmax++;
+        } else if (ISNAN(w)) {
+          // w still within the window and is NA or NaN, x[i] is not NA - already checked above, therefore do nothing
+        } else if (x[i] >= w) {
+          iw = i; w = x[iw];
+        }
+        ans->dbl_v[i] = w;
+      }
+    }
+  }
+  if (verbose)
+    snprintf(end(ans->message[0]), 500, _("%s: nested window max calculation called %"PRIu64" times\n"), __func__, cmax);
+}
+/* fast rolling max - exact
+ * for each observation in x compute max in window from scratch
+ * faster version ignores NAs (narm=T or has.nf=F), as they are not propagated by `>` operator
+ * otherwise we scan for NaN/NA and run either of two loops
+ * has.nf=FALSE can give incorrect results if NAs provided, documented to be used with care
+ */
+void frollmaxExact(double *x, uint64_t nx, ans_t *ans, int k, double fill, bool narm, int hasnf, bool verbose) {
+  if (verbose)
+    snprintf(end(ans->message[0]), 500, _("%s: running in parallel for input length %"PRIu64", window %d, hasnf %d, narm %d\n"), "frollmaxExact", (uint64_t)nx, k, hasnf, (int)narm);
+  for (int i=0; i<k-1; i++) {
+    ans->dbl_v[i] = fill;
+  }
+  if (narm || hasnf==-1) {                                          // ignore NAs as > does not propagate
+    #pragma omp parallel for num_threads(getDTthreads(nx, true))
+    for (uint64_t i=k-1; i<nx; i++) {
+      double w = R_NegInf;
+      for (int j=-k+1; j<=0; j++) {
+        if (x[i+j] > w)
+          w = x[i+j];
+      }
+      ans->dbl_v[i] = w;
+    }
+  } else {
+    bool *isnan = malloc(nx*sizeof(bool));                          // isnan lookup - we use it to reduce ISNAN calls in nested loop
+    if (!isnan) {                                                   // # nocov start
+      ansSetMsg(ans, 3, "%s: Unable to allocate memory for isnan", __func__); // raise error
+      free(isnan);
+      return;
+    }                                                               // # nocov end
+    bool truehasnf = hasnf>0;
+    for (uint64_t i=0; i<nx; i++) {                                 // no openmp as this should be very fast
+      if (ISNAN(x[i])) {
+        truehasnf = true;
+        isnan[i] = true;
+      } else {
+        isnan[i] = false;
```

Comment:
could also calloc and skip assigning the false

Reply:
Could but not sure if we gain much in this. In case of NAs in x we can end up losing some time because they will be first zeroed to then overwritten with ones.